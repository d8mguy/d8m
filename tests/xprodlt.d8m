
// xprodlt: testing a nonstandard $< method
// This triggered a rewrite rule bug so it's in the test suite
val tknclasses = ["ident", "int"]
val testT = extend tuple(tkn: string, xxx: integer) where {
    method $< = \(other: testT) {
        val stkn = self.tkn, otkn = other.tkn
        val seS = (stkn == "stmt" || stkn == "expr"), seO = (otkn == "stmt" || otkn == "expr")
        if(seO) !seS
        else {
            val tclsS = stkn in tknclasses, tclsO = otkn in tknclasses
            if(tclsO) !(seS || tclsS)
            else false
        }
    }
    method $<= = \(other: testT) {
        self < other || self == other
    }
}
val ttoptionStrgs = ["ccc", "ddd", "int", "ident", "expr", "stmt"]
val foo = \() {
    val ttoptions = ttoptionStrgs.{ [testT: this, index ] }
    each(opt1^ttoptions, opt2^ttoptions) println("#{opt1.tkn}, #{opt2.tkn}: #{opt1 < opt2}")
}
foo()
