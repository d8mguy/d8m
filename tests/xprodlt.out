ccc, ccc: false
ccc, ddd: false
ccc, int: true
ccc, ident: true
ccc, expr: true
ccc, stmt: true
ddd, ccc: false
ddd, ddd: false
ddd, int: true
ddd, ident: true
ddd, expr: true
ddd, stmt: true
int, ccc: false
int, ddd: false
int, int: false
int, ident: false
int, expr: true
int, stmt: true
ident, ccc: false
ident, ddd: false
ident, int: false
ident, ident: false
ident, expr: true
ident, stmt: true
expr, ccc: false
expr, ddd: false
expr, int: false
expr, ident: false
expr, expr: false
expr, stmt: false
stmt, ccc: false
stmt, ddd: false
stmt, int: false
stmt, ident: false
stmt, expr: false
stmt, stmt: false
